libs/pageserver_api/src/keyspace.rs

```diff
@@ -14,44 +17,259 @@ pub struct KeySpace {
     pub ranges: Vec<Range<Key>>,
 }
 
+/// Represents a contiguous half-open range of the keyspace, masked according to a particular
+/// ShardNumber's stripes: within this range of keys, only some "belong" to the current
+/// shard.
+///
+/// When we iterate over keys within this object, we will skip any keys that don't belong
+/// to this shard.
+///
+/// The start + end keys may not belong to the shard: these specify where layer files should
+/// start  + end, but we will never actually read/write those keys.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ShardedRange<'a> {
+    pub shard_identity: &'a ShardIdentity,
+    pub range: Range<Key>,
+}
+
+// Calculate the distance between two keys, assuming that they are somewhat close
+// together (i.e. we only account for field5 and field6)
+fn nearby_key_delta(start: &Key, end: &Key) -> u64 {
+    let start = (start.field5 as u64) << 32 | start.field6 as u64;
+    let end = (end.field5 as u64) << 32 | end.field6 as u64;
+    end - start
+}
+
+impl<'a> ShardedRange<'a> {
+    pub fn new(range: Range<Key>, shard_identity: &'a ShardIdentity) -> Self {
+        Self {
+            shard_identity,
+            range,
+        }
+    }
+
+    /// Break up this range into chunks, each of which has at least one local key in it if the
+    /// total range has at least one local key.
+    pub fn fragment(self, target_nblocks: u32) -> Vec<(u32, Range<Key>)> {
+        // Optimization for single-key case (e.g. logical size keys)
+        if self.range.end == self.range.start.add(1) {
+            return vec![(
+                if self.shard_identity.is_key_disposable(&self.range.start) {
+                    0
+                } else {
+                    1
+                },
+                self.range,
+            )];
+        }
+
+        if self.range.end.field1 != self.range.start.field1
+            || self.range.end.field2 != self.range.start.field2
+            || self.range.end.field3 != self.range.start.field3
+            || self.range.end.field4 != self.range.start.field4
+        {
+            // Ranges that span relations are not fragmented.  We only get these ranges as a result
+            // of operations that act on existing layers, so we trust that the existing range is
+            // reasonably small.
+            return vec![(u32::MAX, self.range)];
+        }
+
+        let mut fragments: Vec<(u32, Range<Key>)> = Vec::new();
+
+        let mut cursor = self.range.start;
+        while cursor < self.range.end {
+            let advance_by = self.advance_to_next_boundary(cursor);
+            let is_fragment_disposable = self.shard_identity.is_key_disposable(&cursor);
```

Comment:
This assumes that: "if the first key in the stripe is diposable, then all keys in the stripe are disposable" (same for non-disposable). Is that true?

Reply:
it is since 2d34025e131713dc1183d9d9fd2e2e3fb4c45831

The previous code wasn't quite specific enough in the checks that the low fields of Key were equal: it permitted field5 to differ.  That was done to accommodate ranges that span the logical size key of one relation and the start of the next relation, but would also permit ranges that spanned some unknown sparse area of the keyspace -- and in that latter case, it is _not_ correct to say that the cursor being disposable means the rest of the range is.

So I added a commit 2d34025e131713dc1183d9d9fd2e2e3fb4c45831 that cleans up and clarifies the concept that the check-low-fields branches were doing: they're checking that the key range we're talking about is contiguous.  With that tightened check in place (it requires field6 to be 0xffffffff at the start of the range if field5 differs), the check that this comment points at becomes valid.

Reply:
Okay, I see how this tightens things up, but I'm still having trouble convincing myself the assumption holds.
Let's say we have a range which starts at the logical size key of a relation fork and that's followed by a disposable relation fork range. Since the logical size key is not disposable, that makes the whole range non-disposable which is not right.

Reply:
The trick is that distance_to_next_boundary is 1 if called for the logical size key (special case in `distance_to_next_boundary`).  So the "stripe" is just that one key, and we'll check for disposability again from the 0th block of the following relation

Reply:
Discussed off-line. It works :brain: